### DISCUSSION TITLE
How do I create stress-free strain using eigenstrain?

### DISCUSSION THREAD

**Discussion by jlinBE** [idaholab/moose#17270] [Q&A Modules: Solid mechanics] [answered]
I am building a simple model for mine simulations.

The workflow goes like this: 
1. start with a uniform mesh with a number of blocks pre-defined; everything has the same Young's modulus at this point
2. apply gravity and solve for steady-state
3. import the result from pervious step, except in one of the pre-defined blocks, reduce Young's modulus by several orders of magnitude. again solve for steady-state, so that the mesh deforms.
4. import the result from previous step, except in the same block as before, increase Young's modulus to simulate backfill.  This results in the block pushing back against neighboring blocks because the initial strain is nonzero, which is not what I want.  Instead I would like to set a reference strain (or eigenstrain), which is equal to the previous strain at steady state, and calculate stress using E*(strain - eigenstrain).

I am stuck at step 4.  I think it should be possible using ComputeEigenstrain or ComputeVariableEigenstrain, but not sure about the syntax.

Any help?

@WilkAndy @Traiwit

**Comment by ngrilli** [comment]
Dear @jlinBE

The reduction of the stiffness tensor is not the same thing as introducing an eigenstrain.
For instance during plastic deformation you can have a large eigenstrain (plastic deformation) but still large stress and stiffness.
Or you can have stiffness reduction without any eigenstrain, for instance in a linear elastic model.

An example of code for stiffness reduction based on the temperature is in my repository:
https://github.com/ngrilli/c_pfor_am/blob/main/src/materials/ComputeElasticityTensorMelting.C

The choice between stiffness degradation and introducing eigenstrain depends on the specific phenomenon.

Best Regards,
Nicolò Grilli
National University of Singapore

**Comment by ngrilli** [comment, marked as answer]
About your point 4)
I suggest you to check for elastic-perfectly plastic material models, which is what you are describing.
It may be already available in the tensor mechanics module.

**Reply by ngrilli** [reply]
IsotropicPlasticityStressUpdate object with zero hardening should do that.
Your elements will deform permanently and will not apply stress on the rest of the geometry at the beginning
of your final step (step 4).

**Reply by Traiwit** [reply]
Hi @ngrilli,

Thank you for your suggestion, I tried to set zero hardening at the backfill block, but the simulation does not seem to converge. Not sure if I did anything wrong.

```
  [./isotropic_plasticity_backfill]
    type = IsotropicPlasticityStressUpdate
    yield_stress = 50.0
    hardening_constant = 0
    block = '1'
    use_displaced_mesh = true
  [../]
  [./return_stress_backfill]
  type = ComputeMultipleInelasticStress
     # tangent_operator = elastic
     inelastic_models = 'isotropic_plasticity_backfill'
     block = '1'
  [../]
```

I also attempted @WilkAndy suggestion from #16387: "At the start of each timestep, set the strain in certain elements (block ID) to zero. The "certain elements" are defined by another time-dependent Function, etc. Zeroing the strain should not impact the positions of the nodes in any element. Its purpose is to ensure the stress in these elements is zero (at the start of the timestep). This means an alternative would be to introduce a new reference strain, strain_ref, which is the strain in those elements at the start of the timestep, and calculate stress using stress = E * (strain - strain_ref). You haven't be able to achieve this, since there appears to be nothing in MOOSE that enables either of these ideas."

So for stress = E * (strain - strain_ref), I managed to maintain the deformed mesh using negative values of strain_ref in ComputeEigenstrain (https://mooseframework.inl.gov/source/materials/CompositeEigenstrain.html).

```
[./eigen_backfill]
    type = ComputeEigenstrain
    eigenstrain_name = eigen_true
    eigen_base = '-3e-3 0 0 0 -0 0 0 0 -3e-3' 
    block = '1'
    use_displaced_mesh = true
  [../]
  [./eigen_therest]
    type = ComputeEigenstrain
    eigenstrain_name = eigen_true
    eigen_base = '0'
    block = '0'
    use_displaced_mesh = true
  [../]
```

-3e-3 is obtained from a trial and error method.

The question now is: how do we obtain the exact value of "strain_ref" so we can make sure that the backfill material doesn't push back the deformed mesh. should it be the value of the difference of strain before and after backfilling at that particular block? if so, how do I obtain such a value from 2 output files.


also 1 quick question, why doesn’t moose read second-order AuxVariables (monomial family) from an input file (previous simulation step) when using 
```
  [./e11_aux]
    order = SECOND
    family = MONOMIAL
      initial_from_file_var = e11_aux
  [../]
```

it works fine when the order is constant.

Again, thank you for your time guys :+1:

**Comment by ngrilli** [comment]
Dear @Traiwit 

It's difficult to say what is your convergence problem.
Could you share the complete input file?
The first thing to check for the convergence problem in plasticity is the time step:
an elasto-plastic stress-strain curve must be solved with enough time steps,
maybe you can start with about 1000 time steps to reach 1% total strain or so.
Boundary conditions may also affect.

I suggest you start from the setting of this example with small strain:
https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/test/tests/strain_energy_density/rate_incr_model_elas_plas.i

Best Regards,
Nicolò Grilli
National University of Singapore

**Comment by Traiwit** [comment]
To simplify the question, is there a way that I can change the elasitcity_tensor (E and passion ration) of a block without changing the disp_x,y,z of that block from the previous step.

Thanks guys!
